File: ruby/chapter_1/check_permutation.rb
# @example
#   check_permutation("Tom Marvolo Riddle", "Immortal Love Rodd") #=> true
#   check_permutation("Sam Chong Tay", "Mach Go Nasty") #=> true
#   check_permutation("Tommy", "Tammy") #=> false
def check_permutation(s1, s2)
  # Only keep the relevant chars
  s1 = s1.downcase.delete(' ')
  s2 = s2.downcase.delete(' ')

  # Build hashmap of s2 char counts
  char_count = Hash.new
  char_count.default = 0
  s2.each_char { |c| char_count[c] += 1 }

  # Decrement the hashmap with s1 char counts
  s1.each_char { |c| char_count[c] -= 1 }

  # Return whether or not hashmap is empty
  return char_count.all? { |_, ct| ct == 0 }
end

File: python/chapter_1/check_permutation.py
def check_permutation(s1, s2):
    """Return whether or not s1 is a permutation of s2.

    Ignore spaces for fun testing.

    Uses hash table of character counts:
        - O(max(n,m)) amortized
        - O(nm) worst case

    >>> check_permutation("Tom Marvolo Riddle", "Immortal Love Rodd")
    True
    >>> check_permutation("Tom Marvolo Riddle", "Marmot Dildo Lover")
    True
    >>> check_permutation("Tom Marvolo Riddle", "Mild Doormat Lover")
    True
    >>> check_permutation("Sam Chong Tay", "Mach Go Nasty")
    True
    >>> check_permutation("Tommy", "Tammy")
    False
    """
    # not required by problem, so don't count in complexity analysis
    s1 = ''.join(s1.lower().split())
    s2 = ''.join(s2.lower().split())

    # O(m)
    # create a hash table with counts
    chars = {}
    for c in s2: chars[c] = 1 + chars.get(c, 0)

    # O(n)  amortized
    # O(nm) worst
    # subtract 1 for each c in s1
    for c in s1:
        ct = chars.get(c, 0)
        if ct == 0: return False
        chars[c] = ct - 1

    # O(m)
    return all(ct == 0 for ct in chars.values())

def check_permutation2(s1, s2):
    """More time, less space

    Sorts arrays and then compares them:
        - O(max(n log n, m log m))

    >>> check_permutation("Sam Chong Tay", "Mach Go Nasty")
    True
    >>> check_permutation("Tommy", "Tammy")
    False
    """
    # not required by problem, so don't count in complexity analysis
    s1 = ''.join(s1.lower().split())
    s2 = ''.join(s2.lower().split())

    # check lengths equal first
    if len(s1) != len(s2): return False

    # O (n log n)
    s1 = sorted(s1)
    # O (m log m)
    s2 = sorted(s2)

    # O(max(n, m))
    return s1 == s2
